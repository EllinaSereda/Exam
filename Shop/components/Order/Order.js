import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import './Order.css'
import { voteEvents } from '../../events';

class Order extends React.PureComponent {

  static propTypes = {
    products:PropTypes.array,
    user:PropTypes.shape({
      info:PropTypes.object,
    }),
  } 

  state={
    info:null,
    bigImg:0,
    content:0,
    upd:0,
    del:null,
  }
  

  animateDelete=(EO)=>{
    this.setState({del:EO.target.id});
    setTimeout(this.deleteProd,300);
  }
  deleteProd=(EO)=>{
    let prod=this.state.del.split(' ');
    this.setState({del:prod});
    voteEvents.emit('Delete',prod);
  }

    //  обработчик кнопки -
    dec=(EO)=>{   
      let newState=this.props.user.order;
      let params=EO.target.id.split(' ');
      let prod=this.props.products.filter(v=>v.code==params[0])[0].stock.filter(v=>v.vol==params[1])[0];
      if(1<newState[params[0]][params[1]]){
        newState={...newState,
            [params[0]]:{...newState[params[0]],
              [params[1]]:(newState[params[0]][params[1]]-1),
            }
        }
      }
      voteEvents.emit('upd',newState);
    }
   //  обработчик кнопки +
    inc=(EO)=>{
      let newState=this.props.user.order;
      let params=EO.target.id.split(' ');
      let prod=this.props.products.filter(v=>v.code==params[0])[0].stock.filter(v=>v.vol==params[1])[0];
      if(prod.in>newState[params[0]][params[1]]){
        newState={...newState,
            [params[0]]:{...newState[params[0]],
              [params[1]]:(newState[params[0]][params[1]]+1),
            }
        }
      }
      voteEvents.emit('upd',newState);
    }
   
   
    setAmount=(EO)=>{
      let newState=this.props.user.order;
      let params=EO.target.id.split(' ');
      let prod=this.props.products.filter(v=>v.code==params[0])[0].stock.filter(v=>v.vol==params[1])[0];
      if(prod.in>=EO.target.value){
        newState={...newState,
            [params[0]]:{...newState[params[0]],
              [params[1]]:Number(EO.target.value),
            }
        }
      }
      voteEvents.emit('upd',newState);
    }

  createProductsCode=()=>{
    let prodCode=null;
      prodCode=[];
      for (let key in this.props.user.order){
          let order=[];
          for (let k in this.props.user.order[key]){
             order.push({vol:k,amount:this.props.user.order[key][k]})
          }
          let prod=this.props.products.filter(v=>v.code==key)[0];
          
          prod=order.map((v)=>{
            prod.key=key+' '+v.vol;
            return <div key={prod.key} className={"Prod"+(this.state.del==prod.key?" del":"") }>
              {prod.stock.filter(y=>y.vol==Number(v.vol)).map((x,i)=>{
                return <div className="Variant">
                {(x.in!=0)?
                  <div className="order comp">
                  <input className="Delete" type="button" id={prod.key} onClick={this.animateDelete} value="Удалить"/>
                    <div className="buttons">
                      <span id={prod.key} onClick={this.dec}>-</span>
                      <input type="number" value={v.amount} min={1} id={prod.key} onChange={this.setAmount}/>
                      <span id={prod.key} onClick={this.inc}>+</span>
                    </div>
                  </div>
                  :<input type="button" className="Message" value="Сообщить о наличии"/>
                } 
               <NavLink key={prod.code} to={"/product/"+prod.code} exact className="PageLink" activeClassName="ActivePageLink"><img src={prod.img[0]}/></NavLink>
              <p>{prod.name}</p>
              <p>{v.vol} мл {prod.stock.filter(x=>x.vol==v.vol)[0].price}</p>
              {(x.in!=0)?
               <div className="order adapt">
               <input className="Delete" type="button" id={prod.key} onClick={this.animateDelete} value="Удалить"/>
                 <div className="buttons">
                   <span id={prod.key} onClick={this.dec}>-</span>
                   <input type="number" value={v.amount} min={1} id={prod.key} onChange={this.setAmount}/>
                   <span id={prod.key} onClick={this.inc}>+</span>
                 </div>
               </div>
                :<input type="button" className="Message  adapt" value="Сообщить о наличии"/>
              } 
              </div>})}
          </div>
          })
          prodCode.push(prod);
          
    }

    return  prodCode;
    
  }
  
  
 

  render() {
    let presense=0;
    if (this.props.user){
      let i=0;
      for (let k in this.props.user.order ){
        i++;
      }
      i>0?presense=1:null;
    }
    return <div className="Order">
    {this.createProductsCode()}
   
    </div>

  }

}



export default Order;


