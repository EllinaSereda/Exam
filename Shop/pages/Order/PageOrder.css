.PageOrder{
    width: 80%;
    margin:20px auto;
}
.PageOrder h1{
    text-transform: uppercase;
    font-size: 24px;
    text-align: center;
    font-weight:bold;
    color:#631B42;
    padding:30px 0;
}
.PageOrder .Variant{
    position: relative;
    height:100%;
    padding:10px;
    margin: auto;
    clear:both;
    text-align: left;
    
}
.PageOrder .adapt{
    display:none;
}
.PageOrder .Variant img{
    float:left;
    width:70px;
    padding:10px;
  
}
.PageOrder .Variant h5{
    font-size: 22px;
    color:rgb(84, 84, 84);
    color:rgba(0, 0, 0, 0.3);
    margin-top:25px;
    margin-bottom:15px;
}
.PageOrder .Prod{
    height:110px;
    border-top:1px solid black;
}

.PageOrder .del{
    opacity: 0;
    height:0px;
    transition-duration: 300ms;
    transition-property: height;
}
.PageOrder .del .Variant{ 
    display:none;
}
.PageOrder h1+.Prod{
    border-top:1px solid white;
}


.PageOrder .comp .order{
    float:right;
    margin:30px 0;
    overflow: hidden;
}
.PageOrder .order .Delete, .PageOrder .Message, .PageOrder .MakeOrder {
    font-family: 'Ubuntu', sans-serif;
    color:white;
    border-radius:10px;
    font-size:18px;
    padding:10px 30px;
}
.PageOrder  .Delete, .PageOrder  .Message {
    float:right;
}
.PageOrder .adapt .Delete, .PageOrder .adapt .Message {
    font-size:35px;
    border-radius: 15px;
    width:250px;
    padding:20px 0;
    margin-right:30px;
}

.PageOrder .MakeOrder{
    margin:100px auto 20px;
    clear:both;
    display:block;
    background-color:rgba(156, 60, 118, 0.7);
}
.PageOrder .MakeOrder:hover{
    background-color:rgba(156, 60, 118, 1);
}
.PageOrder .order .Delete{
    background-color:rgba(156, 60, 118, 0.7);
}
.PageOrder .Message{
    margin:35px 0;
    background-color:rgba(213, 195, 221, 0.7);
    
}
.PageOrder .comp .Delete:hover{
    background-color:rgba(156, 60, 118, 1);
}
.PageOrder .comp .Message:hover{
    background-color:#81597E;
}
.PageOrder  .comp .buttons {
    float:right;
    margin:0 10px;
}
.PageOrder  .adapt .buttons {
    margin:auto;
    padding:0 20px;
    width:240px;
}

.PageOrder .buttons{
    padding:0 10px 0 10px;
    border-bottom:2px solid #9C3C76;

}
.PageOrder .order .buttons span{
    font-size:30px;
}
.PageOrder .order input[type="number"]{
    border:none;
    outline:none;
    text-align: center;
} 
.PageOrder .comp input[type="number"]{
    width:70px;
    font-size:18px;
    margin:10px 5px 8px 5px;
} 
.PageOrder .adapt input[type="number"]{
    width:160px;
    font-size:30px;
    margin:20px 10px;
} 

/* Убирает стрелочки */
.PageOrder .order input::-webkit-outer-spin-button,
.PageOrder .order input::-webkit-inner-spin-button {
    -webkit-appearance: none;
}

.PageOrder .order input[type='number'] {
    -moz-appearance: textfield;
}
@media screen  and  (max-width:736pt) {
    .PageOrder{
        width:100%;
    }
}
@media screen and (orientation:portrait) and  (max-device-width:736pt) {
    .PageOrder .MakeOrder{
        position:absolute;
        bottom:60px;
        margin:0;
        font-size:35px;
        border-radius: 15px;
        width:100%;
        padding:40px 0;
    }
    .PageOrder .comp {
        display:none;
    }
    .PageOrder .adapt {
        display:block;
        
    }
    .PageOrder .Prod {
        height:360px;
        
    }
    
    .PageOrder .Variant p{
        margin-top:50px;
        font-size:30px;
    }
    .PageOrder .Variant{
        height:300px;
    }
    
    .PageOrder .Variant img{
        width:300px;

    }
    .PageOrder .del{
        opacity: 0;
        height:0px;
        transition-duration: 300ms;
        transition-property: height;
    }
    .PageOrder .del .Variant{ 
        display:none;
    }
    .PageOrder .adapt input[type="number"]{
        box-sizing:border-box;
        width:80%;
        font-size:30px;

    } 
    .PageOrder .order .buttons span{
        font-size:30px;
        box-sizing:border-box;
        width:10%;
    }
}