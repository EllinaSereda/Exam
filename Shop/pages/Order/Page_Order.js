import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {voteEvents} from '../../events';
import { NavLink } from 'react-router-dom';
import firebase from 'firebase';
import {make_order,product_update} from '../../redux/UserAC';
import Order from '../../components/Order/Order';
import './PageOrder.css';

class Page_Order extends React.PureComponent {

  static propTypes = {
    products:PropTypes.shape({
      products:PropTypes.array,
      search:PropTypes.array,
    }),
  user:PropTypes.shape({
      info:PropTypes.object,
    }),
  } 

  state={
    info:null,
    bigImg:0,
    content:0,
    upd:0,
    del:null,
  }
  

  componentDidMount(){
    voteEvents.addListener('Delete',this.deleteProd);  
    voteEvents.addListener('upd',this.upd);     
  }
  componentWillUnmount = () => {
    voteEvents.removeListener('Delete',this.deleteProd);
    voteEvents.removeListener('upd',this.upd);  
  };

  deleteProd=(prod)=>{
    
    this.setState({del:prod});
    if (this.props.user.info!==null){
        let order=this.props.user.info.order;
      delete order[prod[0]][prod[1]];
      let i=0;
      for (let key in order[prod[0]]){
        i++; 
      }
      if (i==0){
        delete order[prod[0]];
      }
 
      this.props.dispatch(product_update(order));
      firebase.firestore().doc('Users/Users').set({
        [this.props.user.info.email]:{...this.props.user.info,order:order},
      })
      
      
    }
    else if(localStorage.parfumShop_busket){
      let storage=JSON.parse(localStorage.parfumShop_busket);
        delete storage[prod[0]][prod[1]];
        let i=0;
        for (let key in storage[prod[0]]){
          i++;
        }
        if (i==0){
          delete storage[prod[0]];
        }
        localStorage.parfumShop_busket=JSON.stringify(storage);
        this.setState((prevState,CurrentState)=>{return {upd:prevState.upd+1}});
    }
  }

    upd=(newState)=>{   
      firebase.firestore().doc('Users/Users').set({
        [this.props.user.info.email]:{...this.props.user.info,order:newState},
      })
      this.props.dispatch(product_update(newState));
    }
   
   
    setAmount=(EO)=>{
      let newState=this.props.user.info.order;
      let params=EO.target.id.split(' ');
      let prod=this.props.products.products.filter(v=>v.code==params[0])[0].stock.filter(v=>v.vol==params[1])[0];
      if(prod.in>=EO.target.value){
        newState={...newState,
            [params[0]]:{...newState[params[0]],
              [params[1]]:Number(EO.target.value),
            }
        }
      }
      this.props.dispatch(product_update(newState));
    }
    buy=()=>{
      firebase.firestore().doc('Users/Users').set({
        [this.props.user.info.email]:{...this.props.user.info,order:{},buy:[...this.props.user.info.buy,this.props.user.info.order]},
      })
      this.props.dispatch(make_order());
    }

  createProductsCode=()=>{
    let info=null;
    if (this.props.user.info!==null){
      info=this.props.user.info;
    }
    else if (localStorage.parfumShop_busket){
      info=JSON.parse(localStorage.parfumShop_busket);
    }
    return info;
  }
  
  
 

  render() {
    let presense=0;
    if (this.props.user.info){
      let i=0;
      for (let k in this.props.user.info.order ){
        i++;
      }
      i>0?presense=1:null;
    }
    return <div className="PageOrder">
    <h1>Корзина</h1>
    {this.createProductsCode()!==null?
    <Order user={this.createProductsCode()} products={this.props.products.products}></Order>:null}
    
    {presense?
     <input type="button" className="MakeOrder" onClick={this.buy} value="Оформить заказ"/>:
     null
    }
    </div>

  }

}

const mapStateToProps = function (state) {
  return {
    products: state.products,   //Товары
    user: state.user, //Юзер
  };
};

export default connect(mapStateToProps)(Page_Order);


