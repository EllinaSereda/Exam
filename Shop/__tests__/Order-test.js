"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import Order from '../components/Order/Order';

test('работа Order', () => {

  // создаём тестовую версию компонента
  const component = renderer.create(
    <Order user={{name:"aa",email:"eee",order:{30:2}}} products={[{brand:"Carol",code:0,img:["#","#"],info:"parfume",name:"Good",sex:"w",stock:[{in:5,price:122,vol:30}]}]}/>
  );

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  
    
});
