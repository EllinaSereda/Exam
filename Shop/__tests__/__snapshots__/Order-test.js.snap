// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа Order 1`] = `
<div
  className="Order"
/>
`;
